planner/core/exhaust_physical_plans.go

```diff
@@ -450,11 +454,10 @@ func (p *LogicalJoin) getHashJoins(prop *property.PhysicalProperty) (joins []Phy
 		}
 	}
 
-	forced = (p.preferJoinType&preferHashJoin > 0) || forceLeftToBuild || forceRightToBuild
-	noHashJoin := (p.preferJoinType & preferNoHashJoin) > 0
-	if !forced && noHashJoin {
+	forced = forceLeftToBuild || forceRightToBuild
+	if !forced && p.shouldSkipHashJoin() {
 		return nil, false
-	} else if forced && noHashJoin {
+	} else if forced && p.shouldSkipHashJoin() {
```

Comment:
Another problem that needs to solve is to avoid `no-plan` error. 
In the case below, after disabling `Hash Join`, the optimizer can't find a plan for this query.
```
mysql> explain select /*+ no_merge_join(t1) */ * from t1, t2 where t1.a=t2.a;
+------------------------------+----------+-----------+---------------+----------------------------------------------+
| id                           | estRows  | task      | access object | operator info                                |
+------------------------------+----------+-----------+---------------+----------------------------------------------+
| HashJoin_8                   | 12487.50 | root      |               | inner join, equal:[eq(test.t1.a, test.t2.a)] |
| ├─TableReader_15(Build)      | 9990.00  | root      |               | data:Selection_14                            |
| │ └─Selection_14             | 9990.00  | cop[tikv] |               | not(isnull(test.t2.a))                       |
| │   └─TableFullScan_13       | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo               |
| └─TableReader_12(Probe)      | 9990.00  | root      |               | data:Selection_11                            |
|   └─Selection_11             | 9990.00  | cop[tikv] |               | not(isnull(test.t1.a))                       |
|     └─TableFullScan_10       | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo               |
+------------------------------+----------+-----------+---------------+----------------------------------------------+
7 rows in set (0.00 sec)

mysql> set tidb_opt_enable_hash_join=off;
Query OK, 0 rows affected (0.01 sec)

mysql> explain select /*+ no_merge_join(t1) */ * from t1, t2 where t1.a=t2.a;
ERROR 1815 (HY000): Internal : Can't find a proper physical plan for this query
```

Can we implement this in `LogicalJoin.exhaustPhysicalPlans` where we can know whether there is any other possible join plans?
Then to implement it in this way:
```
func (p *LogicalJoin) exhaustPhysicalPlans(...) {
    ...
	if noHashJoin && len(otherJoins) == 0 {
        hashJoins := p.getHashJoins(prop)
        warning("no hash join can not be applied")
        return hashJoins
    } else {
        ...
    }
}
```

Reply:
OK, then we'd better add this into the user doc later on just like what we've done for hints (https://docs.pingcap.com/tidb/dev/optimizer-hints#using-hints-causes-the-cant-find-a-proper-physical-plan-for-this-query-error).

Reply:
I am creating another PR for the doc.

Reply:
see https://github.com/pingcap/docs/pull/14905